src/lib/util/opresolv.cpp

```diff
@@ -19,546 +19,465 @@
 		}
 	}
+}
 
-	// we can't have zero length guidelines strings
-	if (s == specification)
-	{
-		return error::SYNTAX;
-	}
 
-	return error::SUCCESS;
+// -------------------------------------------------
+//	dtor
+// -------------------------------------------------
+
+option_resolution::~option_resolution()
+{
 }
 
 
 // -------------------------------------------------
-//  lookup_in_specification
+//	lookup_in_specification
 // -------------------------------------------------
 
-const char *option_resolution::lookup_in_specification(const char *specification, const option_guide *option)
+const char *option_resolution::lookup_in_specification(const char *specification, const option_guide::entry &option)
 {
 	const char *s;
-	s = strchr(specification, option->parameter);
+	s = strchr(specification, option.parameter());
 	return s ? s + 1 : nullptr;
 }
 
 
 // -------------------------------------------------
-//  ctor
+//	set_specification - sets the option specification
+//	and mutates values accordingly
 // -------------------------------------------------
 
-option_resolution::option_resolution(const option_guide *guide, const char *specification)
+void option_resolution::set_specification(const std::string &specification)
 {
-	const option_guide *guide_entry;
-	int option_count;
-	int opt = -1;
-
-	assert(guide);
-
-	// first count the number of options specified in the guide
-	option_count = count_options(guide, specification);
+	for (auto &entry : m_entries)
+	{
+		// find this entry's info in the specification
+		auto entry_specification = lookup_in_specification(specification.c_str(), entry.m_guide_entry);
 
-	// set up the entries list
-	m_specification = specification;
-	m_entries.resize(option_count);
+		// parse this entry's specification (e.g. - set up ranges and defaults)
+		entry.parse_specification(entry_specification);
 
-	// initialize each of the entries
-	opt = 0;
-	guide_entry = guide;
-	while(guide_entry->option_type != OPTIONTYPE_END)
-	{
-		switch(guide_entry->option_type)
+		// is this a range typed that needs to be defaulted for the first time?
+		if (entry.is_pertinent() && entry.is_ranged() && entry.value().empty())
 		{
-		case OPTIONTYPE_INT:
-		case OPTIONTYPE_ENUM_BEGIN:
-		case OPTIONTYPE_STRING:
-			if (lookup_in_specification(specification, guide_entry))
-				m_entries[opt++].guide_entry = guide_entry;
-			break;
-
-		case OPTIONTYPE_ENUM_VALUE:
-			break;
-
-		default:
-			assert(false && "Invalid option type");
-			break;
+			entry.set_value(entry.default_value());
 		}
-		guide_entry++;
 	}
-	assert(opt == option_count);
 }
 
 
 // -------------------------------------------------
-//  dtor
+//	find
 // -------------------------------------------------
 
-option_resolution::~option_resolution()
+option_resolution::entry *option_resolution::find(int parameter)
 {
+	auto iter = std::find_if(
+		m_entries.begin(),
+		m_entries.end(),
+		[&](const entry &e) { return e.parameter() == parameter; });
```

Comment:
Please don't use auto-capture, it can have surprising effects.  Capture what you need explicitly.